plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs

```diff
@@ -777,6 +797,17 @@ compileExpr e = traceCompilation 2 ("Compiling expr:" GHC.<+> GHC.ppr e) $ do
     -- <error func> <overall type> <message>
     GHC.Var (isErrorId -> True) `GHC.App` GHC.Type t `GHC.App` _ ->
       PIR.TyInst annMayInline <$> errorFunc <*> compileTypeNorm t
+    GHC.Var n `GHC.App` GHC.Type ty
+      | GHC.getName n == mkNilOpaqueName -> case ty of
+          GHC.TyConApp tyCon []
+              | tyCon == GHC.integerTyCon || tyCon == builtinIntegerTyCon ->
+                  pure $ PLC.mkConstant annMayInline ([] @Integer)
+              | tyCon == builtinBoolTyCon -> pure $ PLC.mkConstant annMayInline ([] @Bool)
+              | tyCon == builtinDataTyCon -> pure $ PLC.mkConstant annMayInline ([] @PLC.Data)
+          GHC.TyConApp tyCon [GHC.TyConApp tyArg1 [], GHC.TyConApp tyArg2 []]
+              | (tyCon, tyArg1, tyArg2) == (builtinPairTyCon, builtinDataTyCon, builtinDataTyCon) ->
+                  pure $ PLC.mkConstant annMayInline ([] @(PLC.Data, PLC.Data))
+          _ -> throwPlain $ CompilationError "'mkNil' applied to an unknown type"
```

Comment:
Print the type here?

Reply:
I did think of that, but we don't do it in other places, e.g.

```haskell
  GHC.LitString _ -> throwPlain $ UnsupportedError "Literal string (maybe you need to use OverloadedStrings)"
```

```haskell
  index <- case elemIndex dc dcs of
    Just i -> pure i
    Nothing ->
      throwPlain $
        CompilationError "Data constructor not in the type constructor's list of constructors"
```

```haskell
    GHC.LitAlt _ -> throwPlain $ UnsupportedError "Literal case"
```

so I assumed that the context is getting printed anyway and we don't need to print the type manually.